test/unittests/execute_floating_point_test.cpp

```diff
@@ -700,6 +700,89 @@ TEST(execute_floating_point, f64_promote_f32)
+    // https://webassembly.github.io/spec/core/exec/numerics.html#rounding
+    constexpr double f32_limit = 0x1p128;  // 2**128.
+
+    // The lower boundary input value that results in the infinity. The number is midway between
+    // f32_max and f32_limit. For this value rounding prefers infinity, because f32_limit is even.
+    constexpr double lowest_to_inf = (f32_max + f32_limit) / 2;
+    ASSERT_EQ(lowest_to_inf, 0x1.ffffffp127);
+
+    const std::pair<double, float> test_cases[] = {
+        // demote(+-0) = +-0
+        {0.0, 0.0f},
+        {-0.0, -0.0f},
+
+        {1.0, 1.0f},
+        {-1.0, -1.0f},
+        {double{FP32::Limits::lowest()}, FP32::Limits::lowest()},
+        {double{FP32::Limits::max()}, FP32::Limits::max()},
+        {double{FP32::Limits::min()}, FP32::Limits::min()},
+        {double{FP32::Limits::denorm_min()}, FP32::Limits::denorm_min()},
+
+        // Out of range values rounded to max/lowest.
+        {std::nextafter(f32_max, FP64::Limits::infinity()), FP32::Limits::max()},
+        {std::nextafter(double{FP32::Limits::lowest()}, -FP64::Limits::infinity()),
+            FP32::Limits::lowest()},
+
+        {std::nextafter(lowest_to_inf, 0.0), FP32::Limits::max()},
+        {-std::nextafter(lowest_to_inf, 0.0), FP32::Limits::lowest()},
+
+        // The smallest of range values rounded to infinity.
+        {lowest_to_inf, FP32::Limits::infinity()},
+        {-lowest_to_inf, -FP32::Limits::infinity()},
+
+        {std::nextafter(lowest_to_inf, FP64::Limits::infinity()), FP32::Limits::infinity()},
+        {-std::nextafter(lowest_to_inf, FP64::Limits::infinity()), -FP32::Limits::infinity()},
+
+        // float_32(r) = +inf  (if r >= +limit_32)
+        {f32_limit, FP32::Limits::infinity()},
+
+        // float_32(r) = -inf  (if r <= -limit_32)
+        {-f32_limit, -FP32::Limits::infinity()},
+
+        // demote(+-inf) = +-inf
+        {FP64::Limits::infinity(), FP32::Limits::infinity()},
+        {-FP64::Limits::infinity(), -FP32::Limits::infinity()},
+
+        // The canonical NaN must result in canonical NaN (only the top bit set).
+        {FP32::nan(FP32::canon), FP64::nan(FP64::canon)},
+        {-FP32::nan(FP32::canon), -FP64::nan(FP64::canon)},
+    };
+
+    for (const auto& [arg, expected] : test_cases)
+    {
+        EXPECT_THAT(execute(*instance, 0, {arg}), Result(expected)) << arg << " -> " << expected;
+    }
+
+    // Any input NaN other than canonical must result in an arithmetic NaN.
+    EXPECT_THAT(execute(*instance, 0, {FP64::nan(FP64::canon + 1)}), ArithmeticNaN(float{}));
+    EXPECT_THAT(execute(*instance, 0, {-FP64::nan(FP64::canon + 1)}), ArithmeticNaN(float{}));
+    EXPECT_THAT(execute(*instance, 0, {FP64::nan(1)}), ArithmeticNaN(float{}));
+    EXPECT_THAT(execute(*instance, 0, {-FP64::nan(1)}), ArithmeticNaN(float{}));
+
+    EXPECT_THAT(execute(*instance, 0, {0x1.fffffefffffffp0}), Result(0x1.fffffep0f));  // round down
```

Comment:
Also confused, why is this one rounded down, it's closer to `0x1.ffffffp0f`

Reply:
See comment above.